[RDY] More fixes for 0.1

tarruda opened the pull request:
Continuation of #3309

tarruda committed:
shell: Ensure silent bang mappings won't cut output

Setting `msg_didout` after `call_shell` is enough as it will cause
`hit_return_msg()` to print on next line.

Close #3269

justinmk commented on test/functional/shell/bang_filter_spec.lua:
```diff
@@ -0,0 +1,69 @@
+
+  it('#3269 Last line of shell output is not truncated', function()
+    execute('nnoremap \\l :!ls bang_filter_spec<cr>')
+    feed('\\l')
```
Most tests use `[[]]` instead of escaping backslashes.

tarruda replied:
:+1:

justinmk commented on test/functional/shell/bang_filter_spec.lua:
```diff
@@ -0,0 +1,69 @@
+    os.execute('mkdir bang_filter_spec')
+    os.execute('echo f1 > bang_filter_spec/f1')
+    os.execute('echo f2 > bang_filter_spec/f2')
+    os.execute('echo f3 > bang_filter_spec/f3')
```
There is a `helpers.write_file()` function instead of shelling out to `echo`.

tarruda replied:
:+1:

justinmk commented on test/functional/shell/bang_filter_spec.lua:
```diff
@@ -0,0 +1,69 @@
+
+  before_each(function()
+    clear()
+    os.execute('rm -rf bang_filter_spec')
```
`helpers.rmdir()` is a nice robust alternative that will work on Windows as well.

tarruda replied:
:+1:

justinmk commented on src/nvim/tui/tui.c:
```diff
@@ -677,7 +677,7 @@ static void update_size(UI *ui)
 
   // 2 - try from a system call(ioctl/TIOCGWINSZ on unix)
-  if (!uv_tty_get_winsize(&data->output_handle, &width, &height)) {
+  if (uv_tty_get_winsize(&data->output_handle, &width, &height)) {
```
This doesn't fall through to `$LINES/$COLUMNS` query below on failure.

tarruda replied:
:+1:

tarruda committed:
tui: Don't use 'pastetoggle' for automatic pasting

Add a new special key that can be used by UIs to toggle the 'paste' option and
use it in the TUI instead of the user's 'pastetoggle' value.

Close #2843 #2092

tarruda committed:
os/input: Don't advance past incomplete sequences in input_enqueue

This allows callers to incrementally process buffers that are filled by
incomplete chunks more easily.

tarruda committed:
rbuffer: Enhance rbuffer_reset to work with filled RBuffers

tarruda committed:
tui: Fix queueing of large chunks of input

Take the return value of input_enqueue into consideration, waiting for the main
thread to consume input. Close #1714 #3377.

tarruda committed:
ex_docmd: Fix terminal mode check condition for ex_normal

Using the `curbuf->terminal` condition alone is wrong since it does not
necessarily mean nvim is in terminal mode. It needs to be used with
`State & TERM_FOCUS` because the current buffer could have changed without
`terminal_enter` exiting.

Close #3403

tarruda committed:
test: Add basic tests for the TUI

The tests use `termopen` to spawn nvim and verify the TUI.

justinmk commented:
Test failure on master:

```
Error -> ...uild/neovim/neovim/test/functional/terminal/tui_spec.lua @ 87
tui can handle arbitrarily long bursts of input
./test/functional/ui/screen.lua:278: Row 1 didn't match.
Expected: "item 2997                                         "
Actual:   "{1: }                                                 "
stack traceback:
	./test/functional/ui/screen.lua:278: in function 'wait'
	./test/functional/ui/screen.lua:221: in function 'expect'
	...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:96: in function <...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:87>
```

tarruda commented:
> Test failure on master:

This is because ASAN seems to slow the paste a lot more. I have restarted the build but I think we need to increase the maximum timeout for this test.
